src/Symfony/Component/Message/Transport/Serialization/EncoderInterface.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Message\Transport\Serialization;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+interface EncoderInterface
+{
+    /**
+     * Encode a message to a common format understandable by adapters. The encoded array should only
+     * contain scalar and arrays.
+     *
+     * The most common keys of the encoded array are:
+     * - `body` (string) - the message body
+     * - `headers` (string<string>) - a key/value pair of headers
+     *
+     * @param object $message
+     *
+     * @return array
```

Comment:
What about `Transport\EncodedMessage`?

Reply:
What would be the value?

Reply:
type info :) i.e. not having to rely on such a comment;

```
* The most common keys of the encoded array are:
* - `body` (string) - the message body
* - `headers` (string<string>) - a key/value pair of headers
```

Reply:
But much less flexibility for use-cases we don't know about yet. That's for the same reason that the serializer's context is an array.

Reply:
I've been back and forth on this. 
Serializers are never used in Symfony (message component) code. They are only used by queues to transform a PHP object to array to json. Which means that we should not really care. =)

Except for the `SymfonySerialization` that ships with the component, any implementation of `EncoderInterface` and `DecoderInterface` will only be used inside the concept of a "my-queue-bridge". Which leads to the question, do we need these interfaces?

Reply:
> Serializers are never used in Symfony (message component) code. They are only used by queues to transform a PHP object to array to json.

I've read it this way:

> **Encoders** are never used in Symfony (message component) code. They are only used by **adapters** to transform a PHP object to **a PHP array**. Adapters will then transform the **PHP array** to their own transport layer (can be JSON, can be whatever...)

> Which means that we should not really care. =)

We, as users, that's correct, we shouldn't care at all about these technical details when dispatching messages to the bus. 

> Any implementation of EncoderInterface and DecoderInterface will only be used inside the concept of a "my-queue-bridge".

I don't think so, that's why I've put these interfaces within the component. There are multiple use-cases that are not adapter-specific but might provider encoder/decoders:
- When dispatching messages, it's pretty useful to be able to trace them. To get things like OpenTracing or Zipkin to work well, all we need is "just" proper headers on the messages (via AMQP, HTTP, whatever). So a `symfony-message-zipkin` package could offer encoder/decoders decorators to populate and read these headers, regardless of the transport.
- It might be useful for some users/companies to add their own routing specific keys (in body, header or anything else) within the message before it being sent to either of the senders (which could be different adapters). Such encoder/decoder would help them here.

> Which leads to the question, do we need these interfaces?

The reasons given above expresses my answer which is: I believe we need them. Also, it reduces the duplication within the various adapters and reduces the overhead of creating one.